influxdb3_wal/src/object_store.rs

```diff
@@ -0,0 +1,907 @@
+}
+
+impl WalObjectStore {
+    /// Creates a new WAL. Note that you will have to call replay and then start the flusher
+    /// to fully initialize the WAL.
+    pub fn new(
+        object_store: Arc<dyn ObjectStore>,
+        file_notifier: Arc<dyn WalFileNotifier>,
+        config: WalConfig,
+    ) -> Self {
+        Self {
+            object_store,
+            file_notifier,
+            flush_buffer: Mutex::new(FlushBuffer::new(
+                WalBuffer {
+                    is_shutdown: false,
+                    wal_file_sequence_number: Default::default(),
+                    op_limit: config.max_write_buffer_size,
+                    op_count: 0,
+                    database_to_write_batch: Default::default(),
+                    catalog_batches: vec![],
+                    write_op_responses: vec![],
+                },
+                SnapshotTracker::new(config.snapshot_size, config.level_0_duration),
+            )),
+        }
+    }
+
+    /// Loads the WAL files in order from object store, calling the file notifier on each one and
+    /// populating the snapshot tracker with the WAL periods.
+    pub async fn replay(&self) -> crate::Result<()> {
+        let paths = self.load_existing_wal_file_paths().await?;
+
+        for path in paths {
+            let file_bytes = self.object_store.get(&path).await?.bytes().await?;
+            let wal_contents = verify_file_type_and_deserialize(file_bytes)?;
+
+            // add this to the snapshot tracker so we know what to clear out when
+            self.flush_buffer
+                .lock()
+                .await
+                .snapshot_tracker
+                .add_wal_period(WalPeriod::new(
+                    wal_contents.wal_file_number,
+                    Timestamp::new(wal_contents.min_timestamp_ns),
+                    Timestamp::new(wal_contents.max_timestamp_ns),
+                ));
+
+            match wal_contents.snapshot {
+                None => self.file_notifier.notify(wal_contents),
+                Some(snapshot_details) => {
+                    let snapshot_info = {
+                        let mut buffer = self.flush_buffer.lock().await;
+
+                        match buffer.snapshot_tracker.snapshot() {
+                            None => None,
+                            Some(info) => {
+                                let semaphore = Arc::clone(&buffer.snapshot_semaphore);
+                                let permit = semaphore.acquire_owned().await.unwrap();
+
+                                Some((info, permit))
+                            }
+                        }
+                    };
+
+                    let snapshot_done = self
+                        .file_notifier
+                        .notify_and_snapshot(wal_contents, snapshot_details)
+                        .await;
+                    let details = snapshot_done.await.unwrap();
+                    assert_eq!(snapshot_details, details);
```

Comment:
Good call to change to debug 👍

Reply:
Actually, changed my mind. I do want it to crash, because the next thing it does is delete WAL files. If for some reason the snapshot details we get back aren't expected, we want to crash rather than deleting WAL files, which may be unrecoverable.